src/Symfony/Component/Messenger/Bridge/AmazonSqs/Transport/AmazonSqsReceiver.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Bridge\AmazonSqs\Transport;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\Exception\MessageDecodingFailedException;
+use Symfony\Component\Messenger\Exception\TransportException;
+use Symfony\Component\Messenger\Transport\Receiver\MessageCountAwareInterface;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+use Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface;
+
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class AmazonSqsReceiver implements ReceiverInterface, MessageCountAwareInterface
+{
+    private $connection;
+    private $serializer;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer = null)
+    {
+        $this->connection = $connection;
+        $this->serializer = $serializer ?? new PhpSerializer();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get(): iterable
+    {
+        try {
+            $sqsEnvelope = $this->connection->get();
+        } catch (HttpExceptionInterface $e) {
+            throw new TransportException($e->getMessage(), 0, $e);
+        }
+        if (null === $sqsEnvelope) {
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $sqsEnvelope['body'],
+                'headers' => $sqsEnvelope['headers'],
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            $this->connection->delete($sqsEnvelope['id']);
+
+            throw $exception;
+        }
+
+        yield $envelope->with(new AmazonSqsReceivedStamp($sqsEnvelope['id']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(Envelope $envelope): void
+    {
+        try {
+            $this->connection->delete($this->findSqsReceivedStamp($envelope)->getId());
+        } catch (HttpExceptionInterface $e) {
+            throw new TransportException($e->getMessage(), 0, $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        try {
+            $this->connection->delete($this->findSqsReceivedStamp($envelope)->getId());
```

Comment:
Could you elaborate why `ack` and `reject` both delete the message?

Reply:
SQS does not provides such behavior. It don't care wheter the message has been well handled or not:
It just provide 2 states:
visible = message is available and will be dispatched to the next consumer
invisible = message is currently proceed by a consumer. If the consumer don't delete the message after a TTL, it will automatically return to the "visible" state.

Once a message is proceed  (whether or the result is successfull) and should not be re-dispatched to an oother consumer (retry), it must be deleted.

Reply:
Thank you for clarifying. 

For the record: The `TTL` is the "Visibility timeout". https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html